src/tokenizer.js

```diff
@@ -0,0 +1,792 @@
+// <at-keyword-token>, <hash-token>, <string-token>, <bad-string-token>, <url-token>, <bad-url-token>, <delim-token>,
+// <number-token>, <percentage-token>, <dimension-token>, <whitespace-token>, <CDO-token>, <CDC-token>, <colon-token>,
+// <semicolon-token>, <comma-token>, <[-token>, <]-token>, <(-token>, <)-token>, <{-token>, and <}-token>.
+export class IdentToken extends Token {
+  value;
+  constructor(value) {
+    super();
+    this.value = value;
+  }
+}
+
+export class FunctionToken extends Token {
+  value;
+  constructor(value) {
+    super();
+    this.value = value;
+  }
+}
+
+export class AtKeywordToken extends Token {
+  value;
+  constructor(value) {
+    super();
+    this.value = value;
+  }
+}
+
+export class HashToken extends Token {
+  type;
+  value;
+  constructor(value, type = 'unrestricted') {
+    super();
+    this.value = value;
+    this.type = type;
+  }
+}
+
+export class StringToken extends Token {
+  value;
+  constructor(value) {
+    super();
+    this.value = value;
+  }
+}
+
+export class BadStringToken extends Token {}
+
+export class UrlToken extends Token {
+  value;
+  constructor(value) {
+    super();
+    this.value = value;
+  }
+}
+
+export class BadUrlToken extends Token {}
+
+export class DelimToken extends Token {
+  value;
+  constructor(value) {
+    super();
+    this.value = value;
+  }
+}
+
+export class NumberToken extends Token {
+  value;
+  type;
+  constructor(value, type = "integer") {
+    super();
+    this.value = value;
+    this.type = type;
+  }
+}
+
+export class PercentageToken extends Token {
+  value;
+  constructor(value) {
+    super();
+    this.value = value;
+  }
+}
+
+export class DimensionToken extends Token {
+  value;
+  type;
+  unit;
+  constructor(value, type, unit) {
+    super();
+    this.value = value;
+    this.type = type;
+    this.unit = unit;
+  }
+}
+
+export class WhitespaceToken extends Token {}
+
+export class CDOToken extends Token {}
+
+export class CDCToken extends Token {}
+
+export class ColonToken extends Token {}
+
+export class SemicolonToken extends Token {}
+
+export class CommaToken extends Token {}
+
+export class LeftSquareBracketToken extends Token {}
+
+export class RightSquareBracketToken extends Token {}
+
+export class LeftParenthesisToken extends Token {}
+
+export class RightParenthesisToken extends Token {}
+
+export class LeftCurlyBracketToken extends Token {}
+
+export class RightCurlyBracketToken extends Token {}
+
+class InputStream {
+  codePoints;
+  index = 0;
+  returned = [];
+  constructor(input) {
+    this.codePoints = Array.from(input).map(char => char.codePointAt(0));
+  }
+
+  consume() {
+    if (this.returned.length) {
+      return this.returned.shift();
+    } else {
+      // console.log('Consuming', this.codePoints[this.index]);
+      return this.codePoints[this.index++];
+    }
+  }
+
+  reconsume(codePoint) {
```

Comment:
It looks like reconsume is called with the codepoint that was just consumed. I assume it wouldn't make sense to reconsume something different. If so, can we just step index back instead of needing to maintain a separate returned list that needs to be considered by consume and peek?

Reply:
Yeah.. I was worried that there were situations where a token was reconsumed out of order, but it doesn't seem like it. I've swapped to stepping the index back in https://github.com/flackr/scroll-timeline/pull/206/commits/6ddd897a7b1a9fc62883af224a569d1b1df7b502.

Reply:
Thanks, this looks much simpler now!